lib/block_view.go

```diff
@@ -4726,135 +5590,780 @@ func (bav *UtxoView) _connectUpdateProfile(
-		// them untouched.
-		if len(txMeta.NewUsername) != 0 {
-			newProfileEntry.Username = txMeta.NewUsername
-		}
-		if len(txMeta.NewDescription) != 0 {
-			newProfileEntry.Description = txMeta.NewDescription
-		}
-		if len(txMeta.NewProfilePic) != 0 {
-			newProfileEntry.ProfilePic = txMeta.NewProfilePic
-		}
-		// TODO: Right now a profile can be undeleted by the owner of the profile,
-		// which seems like undesired behavior if a paramUpdater is trying to reduce
-		// spam
-		newProfileEntry.IsHidden = txMeta.IsHidden
+func (bav *UtxoView) _connectNFTBid(
+	txn *MsgBitCloutTxn, txHash *BlockHash, blockHeight uint32, verifySignatures bool) (
+	_totalInput uint64, _totalOutput uint64, _utxoOps []*UtxoOperation, _err error) {
 
-		// Just always set the creator basis points and stake multiple.
-		newProfileEntry.CreatorBasisPoints = txMeta.NewCreatorBasisPoints
+	// Check that the transaction has the right TxnType.
+	if txn.TxnMeta.GetTxnType() != TxnTypeNFTBid {
+		return 0, 0, nil, fmt.Errorf("_connectNFTBid: called with bad TxnType %s",
+			txn.TxnMeta.GetTxnType().String())
+	}
+	txMeta := txn.TxnMeta.(*NFTBidMetadata)
 
-		// TODO: This field is deprecated and should be deleted.
-		newProfileEntry.StakeMultipleBasisPoints = txMeta.NewStakeMultipleBasisPoints
+	// Verify that the postEntry being bid on exists, is an NFT, and supports the given serial #.
+	postEntry := bav.GetPostEntryForPostHash(txMeta.NFTPostHash)
+	if postEntry == nil || postEntry.isDeleted {
+		return 0, 0, nil, RuleErrorNFTBidOnNonExistentPost
+	} else if !postEntry.IsNFT {
+		return 0, 0, nil, RuleErrorNFTBidOnPostThatIsNotAnNFT
+	} else if txMeta.SerialNumber > postEntry.NumNFTCopies {
+		return 0, 0, nil, RuleErrorNFTBidOnInvalidSerialNumber
+	}
 
-		// The StakeEntry is always left unmodified here.
+	// Validate the nftEntry.  Note that there is a special case where a bidder can submit a bid
+	// on SerialNumber zero.  This acts as a blanket bid on any serial number version of this NFT
+	// As a result, the nftEntry will be nil and should not be validated.
+	nftKey := MakeNFTKey(txMeta.NFTPostHash, txMeta.SerialNumber)
+	nftEntry := bav.GetNFTEntryForNFTKey(&nftKey)
+	bidderPKID := bav.GetPKIDForPublicKey(txn.PublicKey)
+	if bidderPKID == nil || bidderPKID.isDeleted {
+		return 0, 0, nil, fmt.Errorf("_connectNFTBid: PKID for bidder public key %v doesn't exist; this should never happen", string(txn.PublicKey))
+	}
 
-	} else {
-		// When there's no pre-existing profile entry we need to do more
-		// checks.
-		if len(txMeta.NewUsername) == 0 {
-			return 0, 0, nil, RuleErrorProfileUsernameTooShort
+	// Save a copy of the bid entry so that we can use it in the disconnect.
+	nftBidKey := MakeNFTBidKey(bidderPKID.PKID, txMeta.NFTPostHash, txMeta.SerialNumber)
+	prevNFTBidEntry := bav.GetNFTBidEntryForNFTBidKey(&nftBidKey)
+
+	if txMeta.SerialNumber != uint64(0) {
+		// Verify the NFT entry that is being bid on exists.
+		if nftEntry == nil || nftEntry.isDeleted {
```

Comment:
I think you want this check outside of the if statement, since bidding on a nil or deleted NFT is always bad.

Reply:
Ah. It's because SerialNumber zero will return a nil NFT. Got it.